### PR TITLE
VAMP

### PR DESCRIPTION
**Pull request by vitskvara**
As described in https://arxiv.org/abs/1705.07120

### CODE REVIEW

**Comment by vitskvara** [comment]
This is the final version. I think the only change that is outside of the scope of the new VAMP structure is the fact that the type signature of `VAE` has changed from 
```julia
VAE{P<:Gaussian,E<:ACD,D<:ACD} <: AbstractVAE
```
to
```julia
struct VAE{P<:CMD,E<:ACD,D<:ACD} <: AbstractVAE
```
since we need VAMP to be a prior of VAE.

File: test/models/vamp.jl [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    model_params_init = get_params(m)
+    @test length(model_params_init) == 9
+
+    # test trainability
+	k = IMQKernel{Float32}(0.1)    
+    λ = .1f0
+    lossf(x) = λ*mmd_mean_vamp(m, x, k) .- mean(logpdf(m.decoder, x, rand(m.encoder, x)))
+    opt = ADAM(0.0001)
+	data = [(x,) for _ in 1:2];
+    Flux.train!(lossf, params(m), data, opt)
```
Comment (nmheim) [inline test/models/vamp.jl:52]: sorry for being a nitpick, but could you format the lines on one consistent indent? :P

File: src/models/vamp.jl [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+"""
+	VAMP{K<:Int,P<:AbstractArray}(K::Int, xdim::Union{Tuple, Int})
+	VAMP{K<:Int,P<:AbstractArray}(X::P)
+
+Vamp prior with `K` pseudoinputs `P`. If only the pseudoinput `X` is given, `K` is the size of 
+the last dimension. 
+
+# Arguments
+* `K`: Number of pseudoinputs
+* `xdim`: Size of an individual pseudoinput
```
Comment (nmheim) [inline src/models/vamp.jl:12]: is xdim used somehwere?

File: src/models/vamp.jl [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+K: 2
+pseudoinputs: Float32[0.876433 0.56233144; 0.24732344 0.44447201]
+)
+
+julia> v = VAMP(2, (3, 2))
+VAMP{Int64,Array{Float32,3}}(
+K: 2
+pseudoinputs: Float32[0.21458945 0.8710681; 0.2402707 -0.20706704; 0.9916858 0.6571086]
+
+Float32[0.6324719 0.27955383; -0.39046887 0.18451884; -0.18170312 -1.3017029]
```
Comment (nmheim) [inline src/models/vamp.jl:29]: is that an overloaded show function? or does julia do that automatically by now?

**Review by nmheim — CHANGES_REQUESTED** [review]
awesome, looks great! just fixing the formatting in the test would be great :)

File: test/models/vamp.jl (hunk 1 above)
Comment (vitskvara) [reply test/models/vamp.jl:52]: Yes, will do :)

File: src/models/vamp.jl (hunk 3 above)
Comment (vitskvara) [reply src/models/vamp.jl:29]: This is automatically generated by Julia, maybe a custom show function would be nice.

File: src/models/vamp.jl (hunk 2 above)
Comment (vitskvara) [reply src/models/vamp.jl:12]: Yes, in the constructor where you only supply the size of a single pseudoinput `xdim` and the number of them that you want to use `K`.

Comment (nmheim) [reply src/models/vamp.jl:12]: ahh, yes I see it now :)

File: src/models/vamp.jl [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+)
+```
+"""
+struct VAMP{K<:Int,P<:AbstractArray} <: CMD # add type here
+	K::K
+	pseudoinputs::P
+
+	VAMP(K::Int, xdim::Union{Tuple, Int}) = new{Int, typeof(randn(Float32, xdim..., K))}(K, randn(Float32, xdim..., K))
+	VAMP(X::AbstractArray) = new{Int, typeof(X)}(size(X)[end], X)
+	VAMP(K::Int, X::AbstractArray) = new{Int, typeof(X)}(K, X) # this is needed for gpu conversion
```
Comment (nmheim) [inline src/models/vamp.jl]: I think defining those as outer constructors could get rid of the special GPU constructor, but I might be mistaken.

Comment (vitskvara) [reply src/models/vamp.jl]: Yeah you are right.

**Review by nmheim — APPROVED** [review]

**Pull request merged**